README
CVSMiner

CVSMiner is suite of programs for automatically mining CVS repositories for
useful information.

CVSMiner requires Python 2.3 or greater.  It uses os.walk which was not
present in Python 2.2.  A workaround may be coming soon.

Copyright (c) 2005 Patrick Wagstrom (pwagstro@andrew.cmu.edu)

src/rcsparse.py is under the terms of a different license as shown
below.  Yes, I can redistribut it, no it's not under terms of the GPL.

Copyright © 1999-2002 The ViewCVS Group. All rights reserved.

By using ViewCVS, you agree to the terms and conditions set forth below:

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

  1. Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in the
     documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND ANY
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.



ChangeLog
2008-11-17  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/eclipse_dashboard_nets/eclipse_nets.R: updated to support
	analysis of the GNOME data set.  It's good enough for now, but not
	quite 100% perfect.
	
	* src/eclipse_dashboard_analysis.py (DashboardNetworkMaker): added support to
	filter the overall project set

2008-11-16  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/eclipse_dashboard_analysis.py: now properly loads in GNOME data too
	* src/dbobjects.py (ProjectInvolvement): added linesRemoved and linesDelta
	* sql/person.sql (user_corporation_project): added lines_removed and lines_delta

2008-11-15  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/eclipse_dashboard_analysis.py: more comments, work on making it
	a bit more generic so it will work with GNOME data too
	* src/dbobjects.py (MasterProject): added in code to link to the parent project if necessary.

2008-11-03  Patirck Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/baseball/boto/awssensitivity.R: updated and added a few comments into the
	code about what I did.

2008-10-20  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>

	* src/baseball/congruence.R: lots of periodic changes update.
	Renamed many of the columns to have them make a lot more sense.

2008-08-11  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/bugs/buglinker.py: script to parse CVS commits looking for bug references

	* sql/bugzilla_additional.sql: added in bz_bugs_cvs_commit to link
	bz_bugs to cvs_commits
	* sql/bugzilla.sql: added in bz_bugs_original_id_idx to index original ids
	on bz_bugs

2008-07-15  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/eclipse_dashboard_nets/eclipse_nets.R: R stats file for analyzing network data

2008-06-24  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/eclipse_dashboard_analysis.py: new file to analyze data obtained from

2008-06-23  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/eclipse_dashboard.py: pulls down contribution information from eclipse.org

	* src/baseball/boto/unpacker.py: script to intelligently unpack all of the data

2008-06-19  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/baseball/boto: updated so it works now.  AWESOME!

2008-06-18  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/baseball/boto: new scripts to farm R processing out to amazon EC2

2008-06-10  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/bugs/eclipseImporter.py: works fine now
	* src/bugs/pgtools.py: helper script for importing eclipse data

2008-06-02  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
	* src/bugs/eclipseImporter.py: fixed bug with null columns

	* src/bugs/eclipseImporter.py: now imports up to longdescs
	* sql/bugzilla.sql: lots of changes to make it work with multiple datasets

2008-05-04  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
    * sql/bugzilla.sql (bz_bugs): creating a new id column instead of the old bug_id column.  Necessary for loading in multiple bugzilla data segments
    
2008-04-14  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>

	* src/dbobjects.py (raw_execute): added new function for database commands
	* src/baseball/bugdumper.py (dump_project_bugs): turn off
	sequential scanning to get a performance boost
	
	* src/baseball/bugdumper.py (process_bug): better work to ensure
	catchings bugs and commits

	* src/baseball/weightednets.py: new functionality to dump bug
	networks
	* src/baseball/weighted.R: massive changes to make it functional

2008-04-11  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
    
    * src/baseball/weightednets.py: added file_id and person_id to node names.  Needed
    for the calculation within R
    
    * src/baseball/weighted.R: initial revision of file 

    * src/baseball/netbreaker.py: script to break apart DynetML output from weightednets
    * src/baseball/netbreaker.ini: config file for netbreaker
    
2008-04-10  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>

    * src/mail/loadMailLists.py: many changes to improve unicode support
    
    * sql/stored_procedures.sql: new file with some helpful stored procedures for
    dealing with mailing list issues.
    
    * src/mail/loadMailLists.py: gave the code a much needed update so it actually
    imports all of the data.
    * src/baseball/weightednets.py: unicode patches
    
2008-04-09  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
    
    * sql/bugzilla_additional.sql: added bz_bugs_product_id_idx
    * src/baseball/weightednets.py: seems to sorta work now
    * src/baseball/weightednets.ini: added lots of new projects to the listing
    
2008-04-08  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>
    
    * src/baseball/weightednets.py: commit file for weighted network analysis
    * src/baseball/weightednets.ini: commit ini file for weighted network analysis
        
    * src/dbobjects.py (raw_query): added new raw_query function to make it easier
    to use raw SQL.
    * src/dbobjects.py: BREAKER merged in all of the stuff from person objects.
    * src/personobjects.py: removed file  
    
2008-04-07  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>

    * src/baseball/congruence.R: finally commited scripts from STC2008

    * sql/drop_eclipse.sql: commented out lines that aren't generally applicable
    
    * src/cvsimporter.py: more cleanup on man errors, handle better now
    
2008-04-06  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>

    * src/bugzillobjects.py: replace StringCol with UnicodeCol
    * src/personobjects.py: replace StringCol with UnicodeCol
    
	* src/cvsimporter.py (get_patch_file, PatchSet.add_line, PatchSet.remove_line):
	deal with silly edge case where csvps replaces "&" with "+++ -"

	* src/cvsimporter.py: check if directory before importing
	* src/cvsimporter_eclipse.py: check if directory before importing

2008-04-05  Patrick Wagstrom  <pwagstro@andrew.cmu.edu>

	* src/drop_eclipse.py: pasted in real instructions at top
	* src/cvsimporter.py (process_patchset): deal with very rare case where a patch
	contains a line of text that looks like the divider between patches    
	

src/cvsimporter.py
#!/usr/bin/python2.4
"""
cvsimporter.py

@author: Patrick Wagstrom
@contact: patrick@wagstrom.net
"""
import subprocess
import datetime
import re
import time
import codecs
import logging
import os
from optparse import OptionParser

MEMBER_RE = re.compile("(.*):([A-Z0-9\.]+)->([0-9\.]+)(\([A-Z]+\))?")

# Constants for parsing patch set information
PATCHSET_SPLIT = "---------------------"
DATE_FORMAT = "%Y/%m/%d %H:%M:%S"
MODE_NORMAL, MODE_MEMBERS, MODE_PATCHES, MODE_LOG = range(1, 5)
TAG_MEMBERS = "Members:"
TAG_PATCH = "Index:"
TAG_PATCH_ALT = "---"

# Set up the logging facilities
logging.basicConfig()
log = logging.getLogger("cvsimporter")

class FileDelta(object):
    """
    Represents a single change on a file.  Used as part of a patch set
    """
    def __init__(self, filename, oldver, newver, state=None):
        object.__init__(self)
        self.filename = filename
        self.oldver = oldver
        self.newver = newver
        self.state = state
        self.linesadded = 0
        self.linesremoved = 0
    
    def add_line(self):
        """
        Increments the counter indicating how many lines this file delta added
        """
        self.linesadded += 1
        
    def remove_line(self):
        """
        Increments the counter indicating how many lines this file delta removed
        """
        self.linesremoved += 1

    def __repr__(self):
        return "FileDelta [%s - added: %d, removed: %d, %s->%s, state: %s]" % (self.filename, self.linesadded, self.linesremoved, self.oldver, self.newver, self.state)

class PatchSet(object):
    """
    Represents a collection of files that were all modified at the same time
    """
    def __init__(self):
        """
        The default constructor doesn't set any properties
        """
        object.__init__(self)
        self.patchid = None
        self.author = None
        self.branch = None
        self.tag = None
        self.log = None
        self.date = None
        self.deltas = {}
        self.text = []
        
    def add_filedelta(self, delta):
        """
        Adds the delta of a single file to the collection of deltas for this
        patch set
        
        @param delta: a FileDelta object
        """
        self.deltas[delta.filename] = delta
        
    def __repr__(self):
        return "PatchSet [id=%s, %d files]\n    " % (self.patchid, len(self.deltas)) + "\n    ".join([str(x) for x in self.deltas.values()])
    
    def set_id(self, val):
        """
        Set the id number for this PatchSet
        
        @param val: an integer representing the value for this patch
        """
        self.patchid = unicode(val)
    
    def set_date(self, val):
        """
        Sets the date for the PatchSet
        
        @param val: a python datetime object
        """
        self.date = val
    
    def set_author(self, val):
        """
        Sets the name the author of this patchset
        
        @param val: the string name of the author
        """
        self.author = unicode(val)

    def set_branch(self, val):
        """
        Sets the branch of the code that this revision was committed to
        
        @param val: the name of the branch, something like HEAD
        """
        self.branch = unicode(val)
        
    def set_tag(self, val):
        """
        Sets the tag applied to this commit operation
        
        Strictly speaking, this is probably not always correct.  Mainly because
        of the fact that tags are sorta strange and may apply to only some files
        that were committed together.
        
        @param val: the name of the tag
        """
        self.tag = unicode(val)
        
    def set_log(self, val):
        """
        
        @param val:
        """
        self.log = unicode(val)

    def add_line(self, filename):
        """
        Informs the patch to add a line to the filedelta for a given file
        
        @param filename: the name of the file
        """
        altfilename = filename.replace("+++ -", "&")
        if self.deltas.has_key(filename):
            self.deltas[filename].add_line()
        elif self.deltas.has_key(altfilename):
            self.deltas[altfilename].add_line()
        else:
            log.warn("patch requested file [%s] -- not found", filename)
    
    def remove_line(self, filename):
        """
        Informs the patch to remove a line to the filedelta for a given file
        
        @param filename: the name of the file
        """
        try:
            self.deltas[filename].remove_line()
        except KeyError:
            log.warn("patch requested file [%s] -- not found", filename)
    
    def append_text(self, newtext):
        """
        The text item keeps track of all the text for a given patchset
        
        @param newtext: the line of text to add
        """
        self.text.append(unicode(newtext)) 
        
    def length(self):
        """
        Gets the lengths of the patchset in lines.  This includes all the other
        assorted stuff that goes with the patchset.
        
        @return: the length of the patch in lines
        @rtype: integer
        """
        return len(self.text)

def import_module(cvsroot, module, diffs=True, fromFile=False, cache=False):
    """
    
    @param cvsroot:
    @param module:
    @param diffs:
    @param fromFile: boolean on if we should force to read from file
    @param cache: boolean if we should try to open 
    """

    filename = "cvsps-%s-%s.txt" % ("_".join([x for x in cvsroot.split("/") if x]), module)
    if cache:
        try:
            open(filename)
            fromFile = True
            log.info("loading from file: %s", filename)
        except IOError:
            log.info("cache file %s not found", filename)
            pass
    if not fromFile:
        # make certain we're working with a directory here
        if not os.path.isdir(cvsroot + os.sep + module):
            raise IOError("Location %s is not a directory")
        if diffs:
            commandstr = ["cvsps", "--cvs-direct", "-g", "--root", cvsroot, module]
        else:
            commandstr = ["cvsps", "--cvs-direct", "--root", cvsroot, module]
        log.debug("executing command: %s", " ".join(commandstr))
        proc = subprocess.Popen(commandstr,
                                stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True, bufsize=1024)
        # trash all that stuff
        proc.stderr.close()
        data = unicode(proc.stdout.read(), encoding="latin-1")
        
        # save a copy of the data
        outfile = codecs.open(filename, "w", "latin-1")
        outfile.write(data)
        outfile.close()
    else:
        log.debug("reading file: %s", filename)
        data = codecs.open(filename, "r", "latin-1").read()
    log.debug("%d bytes of output from cvsps obtained", len(data))
    patchsets = process_data(data)
    return patchsets
    
def parse_member(member, patchset):
    """
    Parses elements of the cvsps output that indicate the members of a patchset.
    
    For example, in the following output, it should create FileDelta objects
    for config and loginfo.
    
    PatchSet 11
    Date: 2001/10/10 20:21:35
    Author: kmoir
    Branch: HEAD
    Tag: (none)
    Log:
    config

    Members:
        config:1.3->1.4
        loginfo:1.5->1.6

    @param patchset: the PatchSet the member is from
    @param member: the line from the output of cvsps like this 
                   "scenarios/PDEScenarios3_3.html:1.2->1.3"
    """
    member = member.strip()
    res = MEMBER_RE.match(member)
    if res:
        filename, originalversion, newversion, newstate = res.groups()
        delta = FileDelta(filename=filename, oldver=originalversion, newver=newversion, state=newstate)
        patchset.add_filedelta(delta)
    else:
        raise Exception("Unable to parse member completely")
        
def process_data(data):
    """
    
    @param data: output from cvsps
    """
    patchsets = []
    currentline = 0
    # split the data if it's just a string
    if data.__class__ == str or data.__class__ == unicode:
        data = data.splitlines()
    while currentline < len(data):
        newpatchset = process_patchset(data[currentline:])
        patchsets.append(newpatchset)
        currentline = currentline + newpatchset.length()
        log.debug("properly imported patchset %s -- up to line %d", newpatchset.patchid, currentline)
    return patchsets
        
def process_patchset(data):
    """
    Processes the data for a single patchset information
    
    This has some interesting peculiarities that arise when parsing.  Namely, the output
    must be exactly that of cvsps -- this means that tabs must remain tabs.
    
    @param data: string of the data from cvsps
    @rtype: a two tuple of a PatchSet and the last line
    """
    if data.__class__ == str or data.__class__ == unicode:
        data = data.splitlines()
    currentpatch = PatchSet()
    currentmode = MODE_NORMAL
    currentlog = ""
    currentfilepatch = None
    for lineno, thisline in enumerate(data):
        currentpatch.append_text(thisline)
        if thisline == PATCHSET_SPLIT:
            if lineno == 0:
                continue
            # this handles a very odd degenerate case...yech
            if currentmode == MODE_PATCHES and len(data) > lineno+1 and not data[lineno+1].startswith("PatchSet"):
                pass
            else:
                return currentpatch
        if currentmode == MODE_NORMAL and thisline.strip():
            try:
                key, val = thisline.strip().split(" ", 1)
            except ValueError:
                key = thisline.strip()
                val = None
            if key == "PatchSet":
                currentpatch.set_id(int(val))
            elif key == "Date:":
                ttime = time.strptime(val, DATE_FORMAT)
                currentpatch.set_date(apply(datetime.datetime, ttime[:7]))
            elif key == "Author:":
                currentpatch.set_author(val)
            elif key == "Branch:":
                currentpatch.set_branch(val)
            elif key == "Tag:":
                if val != "(none)":
                    currentpatch.set_tag(val)
            elif key == "Log:":
                currentmode = MODE_LOG
                currentlog = ""
            elif key == TAG_MEMBERS:
                currentmode = MODE_MEMBERS
            elif key == TAG_PATCH:
                currentmode = MODE_PATCHES
                currentfilepatch = val.split("/", 1)[1]
                patchcounting = False
            elif key == TAG_PATCH_ALT:
                currentmode = MODE_PATCHES
                currentfilepatch = get_patch_file(thisline, data[lineno+1])
                patchcounting = False
            else:
                raise Exception("Unknown key value in parsing line %d: %s" %
                                 (lineno, key))
            
        elif currentmode == MODE_LOG:
            try:
                if data[lineno+1].strip() == TAG_MEMBERS:
                    currentpatch.set_log(currentlog)
                    currentlog = ""
                    currentmode = MODE_NORMAL
                else:
                    currentlog = currentlog + thisline + "\n"
            except:
                raise Exception("end of data stream: " + thisline)
                
        
        elif currentmode == MODE_MEMBERS:
            if not(thisline.strip()):
                currentmode = MODE_NORMAL
            else:
                parse_member(thisline, currentpatch)
            
        elif currentmode == MODE_PATCHES:
            if thisline.startswith("@@"):
                patchcounting = True
            if patchcounting and thisline.startswith("--- ") and len(data) > lineno + 2 and data[lineno+1].startswith("+++ ") and data[lineno+2].startswith("@@ "):
                try:
                    currentfilepatch = get_patch_file(thisline, data[lineno+1])
                    patchcounting = False
                except:
                    print "strange patch formatting on line %d of patch %s" % (lineno, currentpatch.patchid)
            if patchcounting:
                if thisline and thisline[0] == "+":
                    currentpatch.add_line(currentfilepatch)
                if thisline and thisline[0] == "-":
                    currentpatch.remove_line(currentfilepatch)
                if len(data) > lineno + 1 and data[lineno+1].startswith(TAG_PATCH):
                    currentmode = MODE_NORMAL
        
    return currentpatch

def get_patch_file(thisline, nextline):
    """
    Given a patch information header, get the filename we're really concerned about

    Here's an example string to parse:
    
--- pde-ui-home/scenarios/PDEScenarios3_0.html     2008-03-27 17:59:08.785562000 -0400
+++ /dev/null      2004-06-24 14:05:26.000000000 -0400
    
    @param thisline: first line of header
    @param nextline: second line of header
    """

    # fix edge case error where & is replaced with +++ -
    try:
        if thisline.startswith("--- /dev/null"):
            patchfile = nextline.split("\t")[0].split("/", 1)[1]
        else:
            patchfile = thisline.split("\t")[0].split("/", 1)[1]
        return patchfile
    except IndexError:
        raise Exception("l1[%s] - l2[%s]" % (thisline, nextline))

def save_patchset(cvsroot, module, patch):
    """
    saves a given patch to the database
    
    @param patch: a PatchSet object
    """
    print patch


def main_test():
    parser = OptionParser()
    parser.add_option("-l", "--loglevel", dest="loglevel",
                      help="Manually specify logging level (DEBUG, INFO, WARN, etc)",
                      default="INFO", action="store") 
    parser.add_option("-c", "--cvsroot", dest="cvsroot",
                      help="root CVS directory",
                      action="store", default="/home/data/eclipse/cvsroot")
    parser.add_option("-m", "--module", dest="module",
                      help="module to examine",
                      action="store", default="org.eclipse.platform")
    parser.add_option("--usecache", dest="usecache",
                      help="try to use a cache",
                      action="store_true", default=False)
    (options, args) = parser.parse_args() # IGNORE:W0612
    log.setLevel(getattr(logging, options.loglevel.upper()))

    # cvsroot = "/home/data/eclipse/cvsroot/eclipse"
    # module = "org.eclipse.platform"
    patchsets = import_module(options.cvsroot, options.module, cache=options.usecache)
    for pset in patchsets:
        save_patchset(options.cvsroot, options.module, pset)
             
if __name__ == "__main__":
    main_test()
